File: backend/src/invictus_os/services/content_generator.py
from typing import Protocol

from pydantic import ValidationError

from invictus_os.schemas.content import ContentGenerationRequest, GeneratedContentResponse
from invictus_os.services.openai_service import (
    InvalidOpenAIAPIKeyError,
    InvalidOpenAIResponseError,
    MissingOpenAIAPIKeyError,
    OpenAINetworkError,
    OpenAIQuotaError,
    OpenAIRateLimitError,
    OpenAIService,
    OpenAIServiceError,
)


class ContentGenerationError(Exception):
    message = "Content could not be generated right now."


class InvalidContentResponseError(ContentGenerationError):
    message = "OpenAI returned an unexpected content response. Please try generating again."


class ContentGenerator(Protocol):
    def generate(self, request: ContentGenerationRequest) -> GeneratedContentResponse:
        raise NotImplementedError


class OpenAIContentGenerator:
    def __init__(self, *, openai_service: OpenAIService) -> None:
        self._openai = openai_service

    def generate(self, request: ContentGenerationRequest) -> GeneratedContentResponse:
        try:
            payload = self._openai.generate_json(
                system_prompt=content_system_prompt(),
                user_payload={
                    "business_name": request.business_name,
                    "target_audience": request.target_audience,
                    "topic": request.topic,
                    "platform": request.platform,
                    "content_type": request.content_type,
                },
                cache_namespace="content",
            )
            payload["post"] = stringify_content(payload.get("post"))
            payload["reel_script"] = stringify_content(payload.get("reel_script"))
            payload["caption"] = stringify_content(payload.get("caption"))
            payload["call_to_action"] = stringify_content(payload.get("call_to_action"))
            payload["headline"] = stringify_content(payload.get("headline"))
            payload["body"] = stringify_content(payload.get("body"))
            content = GeneratedContentResponse.model_validate(payload)
        except (InvalidOpenAIResponseError, ValidationError, TypeError) as exc:
            raise InvalidContentResponseError from exc
        except (
            MissingOpenAIAPIKeyError,
            InvalidOpenAIAPIKeyError,
            OpenAIRateLimitError,
            OpenAIQuotaError,
            OpenAINetworkError,
        ):
            raise
        except OpenAIServiceError as exc:
            raise ContentGenerationError from exc

        if request.content_type != "reel":
            content.reel_script = None
        elif not content.reel_script:
            raise InvalidContentResponseError

        return content


def stringify_content(value: object) -> str | None:
    if value is None:
        return None

    if isinstance(value, str):
        return value

    if isinstance(value, list):
        sections = [stringify_content(item) for item in value]
        return "\n\n".join(section for section in sections if section)

    if isinstance(value, dict):
        lines: list[str] = []
        for key, item in value.items():
            section = stringify_content(item)
            if section:
                title = str(key).replace("_", " ").title()
                lines.append(f"{title}: {section}")
        return "\n\n".join(lines)

    return str(value)


def content_system_prompt() -> str:
    return (
        "You generate professional healthcare marketing content for InvictusOS. "
        "Write natural, engaging, evidence-informed copy for ethical healthcare promotion. "
        "Do not invent statistics, clinical outcomes, credentials, testimonials, diagnoses, "
        "or treatment guarantees. Avoid fear-based claims and unsafe medical advice. "
        "Maintain Invictus Wellness branding: modern, warm, professional, white-space friendly, "
        "with blue and green wellness positioning. Return only valid JSON with exactly these keys: "
        "headline, body, post, reel_script, caption, hashtags, call_to_action. "
        "headline should be concise and graphic-ready. body should be a clear supporting message. "
        "post should be a complete social media post. caption should be platform-ready. "
        "hashtags must be an array of 5 to 12 relevant strings. call_to_action must be direct. "
        "If content_type is not reel, set reel_script to null. If content_type is reel, include a "
        "complete reel script with hook, scenes, on-screen text, and closing line."
    )

File: frontend/src/services/contentGenerator.ts
import type { ContentGenerationRequest, ContentGenerator, GeneratedContent } from "../types/content";
import { getApiBaseUrl } from "./api";

type ApiGeneratedContent = {
  headline?: string | null;
  body?: string | null;
  post: string;
  reel_script: string | null;
  caption: string;
  hashtags: string[];
  call_to_action: string;
};

const apiBaseUrl = getApiBaseUrl();

function toApiRequest(request: ContentGenerationRequest) {
  return {
    business_name: request.businessName,
    target_audience: request.targetAudience,
    topic: request.topic,
    platform: request.platform,
    content_type: request.contentType,
  };
}

function toGeneratedContent(content: ApiGeneratedContent): GeneratedContent {
  return {
    headline: content.headline ?? undefined,
    body: content.body ?? undefined,
    post: content.post,
    reelScript: content.reel_script ?? undefined,
    caption: content.caption,
    hashtags: content.hashtags,
    callToAction: content.call_to_action,
  };
}

function isApiGeneratedContent(value: unknown): value is ApiGeneratedContent {
  if (!value || typeof value !== "object") {
    return false;
  }

  const content = value as Partial<ApiGeneratedContent>;

  return (
    typeof content.post === "string" &&
    (typeof content.reel_script === "string" || content.reel_script === null) &&
    typeof content.caption === "string" &&
    Array.isArray(content.hashtags) &&
    content.hashtags.every((hashtag) => typeof hashtag === "string") &&
    typeof content.call_to_action === "string"
  );
}

function extractErrorMessage(value: unknown) {
  if (value && typeof value === "object" && "detail" in value) {
    const detail = (value as { detail: unknown }).detail;
    return typeof detail === "string" ? detail : "InvictusOS could not generate content.";
  }

  return "InvictusOS could not generate content.";
}

export class OpenAIContentGenerator implements ContentGenerator {
  async generate(request: ContentGenerationRequest): Promise<GeneratedContent> {
    let response: Response;

    try {
      response = await fetch(`${apiBaseUrl}/content/generate`, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify(toApiRequest(request)),
      });
    } catch {
      throw new Error("InvictusOS could not reach the backend. Confirm the backend is running.");
    }

    const payload: unknown = await response.json().catch(() => null);

    if (!response.ok) {
      throw new Error(extractErrorMessage(payload));
    }

    if (!isApiGeneratedContent(payload)) {
      throw new Error("InvictusOS received an unexpected content response.");
    }

    return toGeneratedContent(payload);
  }
}

export const contentGenerator: ContentGenerator = new OpenAIContentGenerator();
